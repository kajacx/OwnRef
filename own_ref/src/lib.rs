mod macros;
mod structs;

pub use structs::OwnRef;
pub use structs::OwnRefMut;
