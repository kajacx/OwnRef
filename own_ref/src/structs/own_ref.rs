use std::marker::PhantomData;
use std::ops::Deref;
use std::ptr::NonNull;

pub struct OwnRef<'a, T> {
    pointer: NonNull<T>,
    _phantom_lifetime: PhantomData<&'a ()>,
    _phantom_value: PhantomData<T>,
}

impl<'a, T> OwnRef<'a, T> {
    pub(crate) unsafe fn new<'b>(reference: &'b T, _phantom_lifetime: PhantomData<&'a ()>) -> Self {
        Self {
            pointer: NonNull::from(reference),
            _phantom_lifetime,
            _phantom_value: PhantomData,
        }
    }

    pub(crate) fn from_pointer(
        pointer: NonNull<T>,
        _phantom_lifetime: PhantomData<&'a ()>,
    ) -> Self {
        Self {
            pointer,
            _phantom_lifetime,
            _phantom_value: PhantomData,
        }
    }

    pub fn get(&self) -> &T {
        unsafe { self.pointer.as_ref() }
    }

    pub fn take(self) -> T {
        let result = unsafe { std::ptr::read(self.pointer.as_ptr()) };
        std::mem::forget(self); // do not run destructor
        result
    }
}

impl<'a, T> Deref for OwnRef<'a, T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        self.get()
    }
}

impl<T> Drop for OwnRef<'_, T> {
    fn drop(&mut self) {
        let _drop = unsafe { std::ptr::read(self.pointer.as_ptr()) };
    }
}
