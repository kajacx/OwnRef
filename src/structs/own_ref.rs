use std::marker::PhantomData;
use std::ops::Deref;
use std::ptr::NonNull;

pub struct OwnRef<'a, T> {
    pointer: NonNull<T>,
    _phantom_lifetime: PhantomData<&'a ()>,
    _phantom_value: PhantomData<T>,
}

impl<'a, T> OwnRef<'a, T> {
    pub unsafe fn new<'b>(reference: &'b T, _phantom_lifetime: &'a ()) -> Self {
        Self {
            pointer: NonNull::from(reference),
            _phantom_lifetime: PhantomData,
            _phantom_value: PhantomData,
        }
    }

    pub fn get(&self) -> &T {
        unsafe { self.pointer.as_ref() }
    }

    pub fn take(self) -> T {
        let result = unsafe { std::ptr::read(self.pointer.as_ptr()) };
        std::mem::forget(self); // do not run destructor
        result
    }
}

impl<'a, T> Deref for OwnRef<'a, T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        self.get()
    }
}

impl<T> Drop for OwnRef<'_, T> {
    fn drop(&mut self) {
        let _drop = unsafe { std::ptr::read(self.pointer.as_ptr()) };
    }
}
