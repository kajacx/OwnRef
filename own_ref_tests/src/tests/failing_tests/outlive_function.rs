use own_ref::{new_own_ref, OwnRef};

fn main() {
    let text = misbehaving("Hello");

    println!("Oh no, deallocated memory: {}", text.as_str());
}

fn misbehaving(text: &str) -> OwnRef<String> {
    let owned = text.to_owned();

    new_own_ref!(as_own_ref, owned);

    as_own_ref
}
