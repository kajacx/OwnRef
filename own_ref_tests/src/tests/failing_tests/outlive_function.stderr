error[E0515]: cannot return value referencing local variable `_phantom`
  --> src/tests/failing_tests/outlive_function.rs:14:5
   |
12 |     new_own_ref!(as_own_ref, owned);
   |     ------------------------------- `_phantom` is borrowed here
13 |
14 |     as_own_ref
   |     ^^^^^^^^^^ returns a value referencing data owned by the current function
